### ISSUE TITLE
Inefficient String Table Filtering

### ISSUE THREAD

**Issue by astrelsky** [NationalSecurityAgency/ghidra#4783]
**Describe the bug**
Using the filter feature in the string table is inefficient.

**To Reproduce**
Steps to reproduce the behavior: (Task monitor statuses not exact)
1. Open a large program with a lot of strings
2. Open the string table window
3. Copy a string from a different program into the string table filter
4. Remove the end quote from the pasted string
5. Remove the start quote from the pasted string
6. Note the task monitor status as "sorting string table"
7. Note the task monitor status as "filtering string table"
8. Note the task monitor status as "sorting string table"
9. Note the task monitor status as "filtering string table"
10. Note the task monitor status as "sorting string table"
11. Note the task monitor status as "filtering string table"
12. Finally see the filtered results

**Expected behavior**
When the filter is changed the previous filter task should be cancelled and the new one should begin.

**Environment (please complete the following information):**
 - Java Version: Temurin-17.0.4+8
 - Ghidra Version: 10.3_DEV

**Additional context**
You don't need to sort something that is already sorted ;)

**Comment by dragonmacher** [comment]
The table should be responding to the changes in the filter text.   The only minor caveat is that the key presses may be buffered a minor amount to allow the user to press multiple keys before applying the updated filter.

If you are seeing that each change causes a *complete* refilter and sort, then there is presumably an issue with the cancelling of the sort.

> You don't need to sort something that is already sorted ;)

This is perhaps the more interesting part of this ticket.  Currently, each time the data is mutated, we will filter and then sort the remaining filtered items, with the expectation that the filtered subset will require less work to do than the non-filtered data.  (We used to do this in the opposite order.)    Further, as you filter, if your new filter is an extension of the previous filter, then we only filter the last dataset, as opposed to the entire dataset again.   Although, this isn't really relevant to you issue.     

This suggestion here implies an optimization that would take into account the fact that the data was already sorted on the current sort state.    Such a solution would require that we could tell when the original table values remain unmodified.  But, this should be possible.

**Comment by astrelsky** [comment]
I misunderstood what was actually going on. It was the initial sorting when opening the table that was taking forever.

However, is it really necessary to completely redo the sorting and filtering for programActivated events? Constantly switching back and forth between programs while manually comparing and having to wait on the table can be cumbersome.

**Comment by dragonmacher** [comment]
> Just a quick thought. If you filter a sorted list of strings isn't the filtered list already sorted?

Yes, in that case it would be.  The issue for our implementation is that the code doing the filtering/sorting does not know how it was initiated. Thus, the request could be because the data has changed or it could be that the user performed a filter on unchanged data; our code is too course grained in that regard.     We can probably add some extra state to be able to differentiate between the various conditions.

> Constantly switching back and forth between programs while manually comparing and having to wait on the table can be cumbersome.

That is another use case to consider.   Once again, at the point the filtering and sorting is taking place, the plumbing does not know that any previous data was already filtered/sorted.  (Actually, the data is completely reloaded upon switching programs.)    This seems like a common enough use case that it would be nice to have some sort of state to reuse when switching.  But, this could be a major architectural change.

In the meantime, I would suggest opening 2 Code Browsers, one for each program.  Then you can work in either one without paying penalty of switching programs.

**Comment by astrelsky** [comment]
> > Just a quick thought. If you filter a sorted list of strings isn't the filtered list already sorted?
> 
> Yes, in that case it would be.  The issue for our implementation is that the code doing the filtering/sorting does not know how it was initiated. Thus, the request could be because the data has changed or it could be that the user performed a filter on unchanged data; our code is too course grained in that regard.     We can probably add some extra state to be able to differentiate between the various conditions.
> 

Wouldn't the string table be sorted in the database?

**Comment by dragonmacher** [comment]
> Wouldn't the string table be sorted in the database?

Not sure about this particular table, but not all db tables are sorted.  Also, even if sorted, it would only be on one particular set of characteristics, which would not always match the sort on UI table.
